source/common/grpc/async_client_manager_impl.cc

```diff
@@ -33,6 +33,8 @@ AsyncClientFactoryImpl::AsyncClientFactoryImpl(Upstream::ClusterManager& cm,
   if (skip_cluster_check) {
     return;
   }
+  ASSERT(Thread::MainThread::isMainThread(),
+         "async client factory cluster checks should only be performed on the main thread");
```

Comment:
I'm really sorry, I should have looked at this in more detail, but instead of using the `clusters()` api should we just be using the `get()` API which is thread local? This code seems broken. If we did that could we remove the skip clusters check param entirely?

Reply:
I guess we can use the get api, but the skip_clusters_check would become require_static_cluster? I'm not sure what the implications are of a cluster being statically defined vs provided by API or if there are cases that actually require statically defined clusters or if that requirement is due to some legacy behavior that's no longer necessary.

Reply:
actually, I think I can use this: https://github.com/envoyproxy/envoy/blob/main/source/common/config/utility.h#L116..L128

Reply:
Yes ^ exactly, thanks.

/wait

Reply:
the utility function has the same issue, relying on `ClusterManager::clusters()`, so I've just updated this one instance to use `getThreadLocalCluster` and maintain the same semantics as previously for the `skip_cluster_checks` param.

Reply:
The async client factory should only _ever_ be used on the main thread. I think it might be worth asserting that, since you probably will get other kinds of broken behavior if you try use it on a worker thread.

Reply:
This is an example where it's currently being used on a worker thread: https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/ext_authz/config.cc#L80..L89

Does that need to change?

Reply:
Hmm, good point actually, I think what you have makes sense then. CC @chaoqin-li1123 for visibility into this PR.